c/parallel/test/test_for.cpp

```diff
@@ -30,18 +60,62 @@ void for_each(cccl_iterator_t input, uint64_t num_items, cccl_op_t op)
   const char* ctk_path        = TEST_CTK_PATH;
 
   cccl_device_for_build_result_t build;
-  REQUIRE(
-    CUDA_SUCCESS
-    == cccl_device_for_build(&build, input, op, cc_major, cc_minor, cub_path, thrust_path, libcudacxx_path, ctk_path));
+  bool found = false;
+
+  const bool cache_and_key = bool(cache) && bool(lookup_key);
+
+  if (cache_and_key)
+  {
+    auto& cache_v     = cache.value();
+    const auto& key_v = lookup_key.value();
+    if (cache_v.contains(key_v))
+    {
+      build = cache_v.get(key_v).get();
+      found = true;
+    }
+  }
+
+  if (!found)
+  {
+    REQUIRE(CUDA_SUCCESS
+            == cccl_device_for_build(
+              &build, input, op, cc_major, cc_minor, cub_path, thrust_path, libcudacxx_path, ctk_path));
+
+    if (cache_and_key)
+    {
+      auto& cache_v     = cache.value();
+      const auto& key_v = lookup_key.value();
+      cache_v.insert(key_v, build);
+    }
+  }
+
   const std::string sass = inspect_sass(build.cubin, build.cubin_size);
   REQUIRE(sass.find("LDL") == std::string::npos);
   REQUIRE(sass.find("STL") == std::string::npos);
 
   REQUIRE(CUDA_SUCCESS == cccl_device_for(build, input, num_items, op, 0));
-  REQUIRE(CUDA_SUCCESS == cccl_device_for_cleanup(&build));
+  if (cache_and_key)
+  {
+    // if cache and lookup_key were provided, the ownership of resources
+    // allocated for build is transferred to the cache
+  }
```

Comment:
Is this empty on purpose?

Reply:
It is empty on purpose, so that the purpose can be documented in a comment. That is, since the ownership was transferred, we do not need to manually releases resources allocated by building the algorithm,